### PR TITLE
Pipe rpc

### PR DESCRIPTION
**Pull request by froody**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Adds support for:
* Reused layers (e.g. for weight sharing)
* Lazily-constructed layers
* Single-process control via PipeRPCWrapper

Also lays the foundation for asynchronous pipeline work by introducing an event loop for each rank/worker to process either activations or gradients as they arrive.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/pipe.py [hunk 1]
```diff
@@ -11,16 +11,26 @@
 from torch.distributed import rpc
 import torch.multiprocessing as mp
 import torch.nn as nn
+from torch.nn.parallel import DistributedDataParallel as DDP
 from torch.utils.data import DataLoader
 import torchtext
 from torchtext.data.utils import get_tokenizer
 
+# from deepspeed.pipe import PipelineModule
```
Comment (msbaines) [inline benchmarks/pipe.py:19]: Remove.

File: tests/nn/model_parallel/commons.py [hunk 2]
```diff
@@ -35,6 +36,12 @@
 from fairscale.nn.model_parallel import initialize_model_parallel
 from fairscale.nn.model_parallel.random import model_parallel_cuda_manual_seed
 
+try:
+    import torch_ucc  # noqa: F401
```
Comment (msbaines) [inline tests/nn/model_parallel/commons.py:40]: This import seems unused.

File: benchmarks/pipe.py [hunk 3]
```diff
@@ -11,16 +11,26 @@
...
 from fairscale.nn.model_parallel import initialize_model_parallel
-from fairscale.nn.pipe import pipe
+from fairscale.nn.model_parallel.initialize import get_data_parallel_group, get_pipeline_parallel_group
+from fairscale.nn.pipe import LazyModule, pipe
 from fairscale.optim import GradScaler
+from fairscale.optim.oss import OSS
 from tests.nn.model_parallel.commons import dist_init, get_worker_map
 
+try:
+    import torch_ucc  # noqa: F401
```
Comment (msbaines) [inline benchmarks/pipe.py:29]: This import does not seem to be used.

File: benchmarks/pipe.py [hunk 4]
```diff
@@ -276,7 +289,15 @@ def train(lm_dataloader, model, criterion, optimizer, vocab_size, args):
 
     num_params = reduce(operator.add, (reduce(operator.mul, x.size()) for x in model.parameters()))
     if model.group:
-        print(f"training model, #prams = {num_params}, group: {model.group.rank()}, sizes {model.group.size()}")
+        total = torch.Tensor([num_params]).cuda()
+        torch.distributed.all_reduce(total, group=model.group)
+        print(
```
Comment (msbaines) [inline benchmarks/pipe.py:294]: use logging instead of print

File: tests/nn/model_parallel/commons.py [hunk 5]
```diff
@@ -99,10 +113,20 @@ def spawn_for_all_world_sizes(test_func, world_sizes=get_world_sizes(), args=[])
         mp.spawn(test_func, args=(world_size, *args), nprocs=world_size, join=True)
 
 
-def helper(rank, world_size, func, args):
+def helper(rank, world_size, func, args, error_queue):
```
Comment (msbaines) [inline tests/nn/model_parallel/commons.py:116]: Maybe a better name and a comment. Ditto for fixer. The terse names make understanding the code challenging.

File: tests/nn/model_parallel/test_layers.py [hunk 6]
```diff
@@ -307,6 +307,10 @@ def run_test_row_parallel_linear(rank, model_parallel_size):
         print(" >> passed the test :-)")
 
 
+def a_barrier():
```
Comment (msbaines) [inline tests/nn/model_parallel/test_layers.py:310]: The code felt cleaner when it called torch.distributed.barrier() directly.

File: tests/nn/moe/test_moe_layer.py [hunk 7]
```diff
@@ -23,12 +23,14 @@
...
 if "OMPI_COMM_WORLD_SIZE" in os.environ:
-    dist.init_process_group(backend=dist.Backend.MPI)
+    pass  # dist.init_process_group(backend=dist.Backend.MPI)
 
 
 def setup_module(module):
     if "OMPI_COMM_WORLD_SIZE" not in os.environ:
         dist.init_process_group(backend=BACKEND, rank=0, world_size=1)
+    else:
+        dist.init_process_group(backend=dist.Backend.MPI)
```
Comment (msbaines) [inline tests/nn/moe/test_moe_layer.py:33]: This doesn't work. I get an error about MPI being initialized multiple times.

**Review by msbaines — COMMENTED** [review]
Looks good. Mostly nits.

File: tests/nn/moe/test_moe_layer.py (hunk 7 above)
Comment (froody) [reply tests/nn/moe/test_moe_layer.py:33]: ah, fixed, tested locally and it passes

File: fairscale/nn/pipe/messages.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+import numpy as np
+import torch
+from torch import Tensor
+
+from fairscale.nn.model_parallel import get_pipeline_parallel_group
+
+from .types import MESSAGE_GENERATION_START, InputDevice, PipeMessage, Tensors, TransportConfig
+
+# FIXME Why is 256 ok for training but not for tests?
+MESSAGE_TENSOR_SIZE = 1024  # 256
```
Comment (msbaines) [inline fairscale/nn/pipe/messages.py:15]: Why does this need to be fixed-size?

File: fairscale/nn/pipe/messages.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+MessageQueues: List[Queue] = [Queue() for _ in range(MESSAGE_GENERATION_START)]
+
+
+def pyobject_to_tensor(obj: Any) -> Tensor:
```
Comment (msbaines) [inline fairscale/nn/pipe/messages.py:20]: We have similar code in fairscale/optim/utils.py. Maybe we should create a fairscale.utils package to avoid duplication.

File: fairscale/nn/pipe/messages.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    else:
+        return tuple(t.to(input_device) for t in tensors)
+
+
+def rpc_push_queue(message: PipeMessage) -> None:
+    globals()["MessageQueues"][message.queue_name].put(message)
+
+
+def send_message(config: TransportConfig, message: PipeMessage, sync: bool = False) -> None:
+    if config.use_rpc:
```
Comment (msbaines) [inline fairscale/nn/pipe/messages.py:57]: Might be nice to avoid the multiple config.use_rpc checks and just have a separate RPCMessageQueue class.

File: fairscale/nn/pipe/messages.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        return tensors
+    else:
+        return tuple(t.to(input_device) for t in tensors)
+
+
+def rpc_push_queue(message: PipeMessage) -> None:
+    globals()["MessageQueues"][message.queue_name].put(message)
+
+
+def send_message(config: TransportConfig, message: PipeMessage, sync: bool = False) -> None:
```
Comment (msbaines) [inline fairscale/nn/pipe/messages.py:56]: I don't see a test for this or any callers in this PR. Is this dead code or is their a follow-up PR that uses this?

**Review by msbaines — COMMENTED** [review]
If possible it would be nice if this could be broken up. Its hard to review all at once. It would also be useful to have a description of the design in one of the files as a comment or docstring.

File: benchmarks/pipe.py [hunk 12]
```diff
@@ -460,10 +530,13 @@ def bench_single_process(args):
     blob = make_model_and_data(args, None, new_data=new_data)
     model = blob["model"]
 
-    balance = generate_balance(min(num_devices, 8), len(model))
-    p = pipe.Pipe(
-        model, balance, chunks=args.chunks, pipelined_backward=args.pipelined_backward, checkpoint=args.checkpoint
-    )
+    if args.deepspeed:
+        p = PipelineModule(layers=model, num_stages=min(num_devices, 4))
```
[the remark this answers is not in the document]
Comment (froody) [reply benchmarks/pipe.py:534]: hah you told me to delete the import, it never worked though so I'll remove it

File: fairscale/nn/pipe/messages.py (hunk 8 above)
Comment (froody) [reply fairscale/nn/pipe/messages.py:15]: Because send/recv have to agree on the size, so the two options are send the size first, or just pre-allocate a large enough buffer for all cases, so I chose the latter

File: fairscale/nn/pipe/messages.py (hunk 9 above)
Comment (froody) [reply fairscale/nn/pipe/messages.py:20]: oh I didn't know about that, created fairscale/utils/object.py

File: fairscale/nn/pipe/messages.py (hunk 10 above)
Comment (froody) [reply fairscale/nn/pipe/messages.py:57]: good point

File: fairscale/nn/pipe/messages.py (hunk 11 above)
Comment (froody) [reply fairscale/nn/pipe/messages.py:56]: It's called from pipeline.py and async_schedule.py

**Comment by msbaines** [comment]
Added @blefaudeux to review fairscale.utils changs.

File: fairscale/optim/utils.py [hunk 13]
```diff
@@ -52,11 +53,8 @@ def broadcast_object(
 
     if dist.get_rank() == src_rank:
         # Emit data
-        buffer = io.BytesIO()
-        torch.save(obj, buffer)
-        data = bytearray(buffer.getbuffer())
-        length_tensor = torch.LongTensor([len(data)]).to(dist_device)
-        data_send_tensor = torch.ByteTensor(data).to(dist_device)
+        data_send_tensor = pyobject_to_tensor(obj).to(dist_device)
```
Comment (blefaudeux) [inline fairscale/optim/utils.py:56]: ah interesting, I didn't know of this way ! Could you elaborate on the benefits vs. BytesIO ? cc @mannatsingh, because this code comes from Classy initially. Else if that works better in any way that's fine by me, the code is a little cleaner I think

Comment (mannatsingh) [reply fairscale/optim/utils.py:56]: Ah yeah this looks like a function within fairsclae which does something similar to what we do in classy, but through pickle and numpy (https://github.com/facebookresearch/fairscale/blob/63f7796a72f9afbcba64f9bf0df11753ecc4558d/fairscale/nn/pipe/pipeline.py#L108). Although I don't think we're supposed to pickle tensors. So if obj contains tensors this might become problematic?

Comment (froody) [reply fairscale/optim/utils.py:56]: Ok, I'm going to revert my changes to fairscale.optim

Comment (mannatsingh) [reply fairscale/optim/utils.py:56]: Just mentioned this to @blefaudeux over chat - I think the only issue with pickling tensors is the device handling bit. 
So I guess my statement isn't completely true - "I don't think we're supposed to pickle tensors". If the code works with both GPUs and CPUs, there's no need to get rid of that. Just something to be mindful of :)

**Review by msbaines — APPROVED** [review]

**Review by sidgoyal78 — COMMENTED** [review]
Thanks for the PR. Looks great. The modules for async are already being used for ampnet/xpipe.

**Review by sidgoyal78 — APPROVED** [review]
